### PR TITLE
Add APIv4 architecture docs

### PR DESCRIPTION
**Pull request by colemanw**
Fills in missing docs about how to build onto the api.

### CODE REVIEW

File: docs/api/index.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ Utilizing the API is superior to accessing core functions directly (e.g.calling
 
 The best place to begin working with the API is your own *test* install of CiviCRM, using the API explorer and the API parameter list.
 
-Extensions can provide additional API functionality. For help creating your own additions to the API, see [civix generate:api](/extensions/civix.md#generate-api).
+Extensions can provide additional API functionality. For help creating your own additions to the API, see [Api Architecture](/api/v4/architecture.md).
```
Comment (MikeyMJCO) [inline docs/api/index.md]:
```suggestion
Extensions can provide additional API functionality. For help creating your own additions to the API, see [API Architecture](/api/v4/architecture.md).
```

File: docs/api/v4/architecture.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
+# APIv4 Architecture
+
+*This guide is intended for anyone wishing to extend the CiviCRM API itself, either in core or in their own extension.*
+
+## API Entity Classes
+
+Every API entity is a class which inherits from [`\Civi\Api4\Generic\AbstractEntity`]((https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Generic/AbstractEntity.php)). Each class serves two purposes:
+
+1.  Declare the API entity by its existance.
+    - You can declare a new api (e.g. FooBar) simply by placing the class `\Civi\Api4\FooBar` in the `/Civi/Api4` directory of your extension, as long as it inherits from `\Civi\Api4\Generic\AbstractEntity`.
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
    - You can declare a new API (e.g. FooBar) simply by placing the class `\Civi\Api4\FooBar` in the `/Civi/Api4` directory of your extension, as long as it inherits from `\Civi\Api4\Generic\AbstractEntity`.
```

File: docs/api/v4/architecture.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    
+2.  Provide factory functions for that entity's **Actions**.
+
+## API Action Classes
+
+Every API action is a class which inherits from [`\Civi\Api4\Generic\AbstractAction`]((https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Generic/AbstractAction.php)). It has two functions:
+
+1.  Store the parameters of the API call.
+    - Every `protected` class var is considered a parameter (unless it starts with an underscore).
+    - The `$version` parameter refers to the api version (4) and cannot be changed.
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
    - The `$version` parameter refers to the API version (4) and cannot be changed.
```

File: docs/api/v4/architecture.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    - The `$_entityName` and `$_actionName` vars are set when the Action object is constructed (by the Entity class factory fn, see above). The underscore makes them hidden parameters as they cannot be changed. 
+    - Adding a `protected` var to your Action named e.g. `$thing` will automatically:
+        - Provide a getter/setter (via `__call` MagicMethod) named `getThing()` and `setThing()`.
+        - Expose the param in the Api Explorer (be sure to add a doc-block as it displays in the help panel).
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
        - Expose the param in the API Explorer (be sure to add a doc-block as it displays in the help panel).
```

File: docs/api/v4/architecture.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+2.  Define a `_run()` function to execute the call and return the results.
+    - The `_run()` function is invoked by `execute()`
+    - It gets passed a `Result` object into which it places an array of values from whatever logic it executes.
+    
+## The Result Class
+
+An API [`Result`](https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Generic/Result.php) is a PHP ArrayObject. It has three functions:
+
+1.  Store the results of the API call (accessible via ArrayAccess).
+
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
```

File: docs/api/v4/architecture.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+2.  Store metadata like the Entity & Action names.
+    - Some actions extend the Result object to store extra metadata. For example [`BasicReplaceAction`](https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Generic/BasicReplaceAction.php) returns [`\Civi\Api4\Result\ReplaceResult`](https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Result/ReplaceResult.php) which includes the additional `$deleted` property to list any items deleted by the operation.
+
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
```

File: docs/api/v4/architecture.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+}
+```
+
+</div>
+
+---
+
+These are typical CRUD APIs using generic functions to perform actions on the `civicrm_website` table.
+The v3 file needed a function for each action, resulting in a lot of duplicate code across api files.
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
The v3 file needed a function for each action, resulting in a lot of duplicate code across API files.
```

File: docs/api/v4/architecture.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+![Entity Inheritance Diagram](/img/APIv4-entity-inheritance.svg)
+
+There are two categories of APIv4 entities: "standard" (aka DAO) entities, and ad-hoc entities.
+
+- Standard entities correspond to a database table and `DAO` class. E.g. the *Contact* API entity corresponds to the
+`CRM_Conatact_DAO_Contact` entity and the `civicrm_contact` database table. These classes extend from `Civi\Api4\Generic\DAOEntity` and thus inherit the standard set of DAO actions. 
+- Ad-hoc entities extend the [`\Civi\Api4\Generic\AbstractEntity`](https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Generic/AbstractEntity.php) class directly. They have the flexibility to implement any actions, using any datasource.
+  They are not required to implement any action except `GetFields`.
+  See the [Api4 example extension](https://lab.civicrm.org/extensions/api4example/) for a working demonstration.
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
  See the [APIv4 example extension](https://lab.civicrm.org/extensions/api4example/) for a working demonstration.
```

File: docs/api/v4/architecture.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+Each basic action is implemented by supplying a callback function to perform logic and return data,
+and the basic action handles details like validating param input and formatting & filtering the result.
+They are designed to be used in 1 of 2 ways:
+
+1. Your Entity's action function can construct the basic action directly, passing the callback into the constructor.
+2. You can override the basic action class (e.g. for customizing the parameters). See [`Example::create()`](https://lab.civicrm.org/extensions/api4example/blob/master/Civi/Api4/Action/Example/Create.php).
+
+- [**`AbstractAction`**](https://github.com/civicrm/civicrm-core/blob/master/Civi/Api4/Generic/AbstractAction.php):
+  The base class for all api actions. Every action must, at minimum, extend this class.
```
Comment (MikeyMJCO) [inline docs/api/v4/architecture.md]:
```suggestion
  The base class for all API actions. Every action must, at minimum, extend this class.
```

File: docs/api/v4/usage.md [hunk 10]
```diff
@@ -37,6 +37,8 @@ There are two ways to call the api from PHP - which one you choose is a matter o
 For example you may prefer OOP syntax because IDE code editors provide autocompletion.
 Or if you need to work with the parameters as an array, traditional syntax will be more convenient.   
 
+[Api4 PHP Examples](/img/Api4-PHP-Styles.svg)
```
Comment (MikeyMJCO) [inline docs/api/v4/usage.md]:
```suggestion
[APIv4 PHP Examples](/img/Api4-PHP-Styles.svg)
```

**Comment by colemanw** [comment]
Thanks @MikeyMJCO I've applied your suggestions.

**Comment by MikeyMJCO** [comment]
This is merge-ready from me once you agree!

**Pull request merged**